README.md
# The Accedo One Express middleware [![npm](https://img.shields.io/npm/v/@accedo/accedo-one-express.svg?maxAge=3600)](https://www.npmjs.com/package/@accedo/accedo-one-express)

```
   _                    _           ___
  /_\   ___ ___ ___  __| | ___     /___\_ __   ___
 //_\\ / __/ __/ _ \/ _` |/ _ \   //  // '_ \ / _ \
/  _  \ (_| (_|  __/ (_| | (_) | / \_//| | | |  __/
\_/ \_/\___\___\___|\__,_|\___/  \___/ |_| |_|\___|

```

## Summary

This is the official [Accedo One](https://www.accedo.tv/one/) middleware for Express.

The [Accedo One SDK] for Javascript(https://github.com/Accedo-Products/accedo-one-sdk-js/) provides an easy way to make use of Accedo One APIs.
This middleware provides additional benefits and makes things even easier in the context of an Express app.

We follow [semantic versioning](http://semver.org/), and you may have a look at our change log [here](./CHANGELOG.md).

## Compatibility

This middleware is written in ES6 (with CommonJS modules rather than ES6 Modules).

This should work from version 4, but we test on, and recommend to use the latest LTS version of Node (currently Node 6, soon Node 8).

For earlier Node versions, you may try introducing a compilation step to ES5 (through Babel or Buble, for instance). Note we did not test this and **strongly suggest not using a deprecated Node version** as they may lack important security fixes.

Regarding Express, we support and test against version 4.

## Features

Features inherited from the Accedo One SDK for Javascript :
 - easy access to Accedo One APIs
 - automatic deviceId creation when none was provided
 - automatic session creation when none was provided (lazy - only when needed)
 - automatic session re-creation when the existing one has expired (lazy)
 - ensures only one session will be created at a time, even if a request triggers concurrent Accedo One calls

Extra features provided by this Express middleware :
 - automatic creation of Accedo One client instances for each request, attached to the response object for further use
 - automatically passes the requester's IP onto Accedo One calls for analytics and geolocated services
 - automatic reuse of the deviceId through cookies (can be customized to use anything else based on requests)
 - automatic reuse of the sessionKey through cookies (can be customized to use anything else based on requests)

Note when you use the middleware, you should also [configure Express to handle proxies correctly](http://expressjs.com/en/4x/api.html#trust.proxy.options.table) as we rely on the IP it gives us.
Do read the Express doc, and the note on this further below.

For instance: `app.set('trust proxy', ['loopback', 'linklocal', 'uniquelocal'])`

## Documentation

| What | Link
|------|------
| The Accedo One Express middleware (this project) | https://accedo-products.github.io/accedo-one-sdk-express/
| The Accedo One JS SDK | https://accedo-products.github.io/accedo-one-sdk-js/
| The Accedo One REST APIs | https://developer.one.accedo.tv

Each Express response gets associated to an Accedo One client instance, found in `res.locals.accedoOneClient` (when `res` is the response variable name).

To find what methods are available on these instances, refer to the API docs for the Accedo One JS SDK listed above.

The doc for the REST APIs is also listed as Accedo One-specific terminology is defined there.

## Installation

| Your preferred CLI tool | Command
|------|------
|Yarn|`yarn add @accedo/accedo-one @accedo/accedo-one-express`
|NPM|`npm install --save @accedo/accedo-one @accedo/accedo-one-express`

Then you can get the default export to use the middleware:

| Method | Code
|------|------
|CommonJS| <pre lang="js">const accedoOne = require('@accedo/accedo-one-express')</pre>
|ES6 Module| <pre lang="js">import accedoOne from '@accedo/accedo-one-express'</pre>

## Examples

### Use the middleware to persist deviceId and sessionKey via cookies

```js
const accedoOne = require('@accedo/accedo-one-express');
const express = require('express');

const PORT = 3000;

express()
// handle proxy servers if needed, to pass the user's IP instead of the proxy's. Read more about this further.
.set('trust proxy', ['loopback', 'linklocal', 'uniquelocal'])
// place the accedoOne middleware before your request handlers
.use(accedoOne({ appKey: '56ea6a370db1bf032c9df5cb' }))
.get('/test', (req, res) => {
   // access your client instance, it's already linked to the deviceId and sessionKey via cookies
   res.locals.accedoOneClient.getEntryById('56ea7bd6935f75032a2fd431')
   .then(entry => res.send(entry))
   .catch(err => res.status(500).send('Failed to get the result'));
})
.listen(PORT, () => console.log(`Server is on ! Try http://localhost:${PORT}/test`));
```

See also examples in the documentation linked above.

## SDK development

  * Clone/fork this repo
  * Run `yarn` (install it first if needed)
  * Develop !
  * Before pushing, remember to:
    - add tests and check they all pass (`npm test`)
    - document any public API with JSDoc comments and generate the new doc (`npm run doc`)

## More information & Links

* [Accedo One homepage](https://www.accedo.tv/one)
* [Accedo One API documentation](https://developer.one.accedo.tv)

## On the Express `trust proxy` setting, and propagating the right IP

This middleware will set the incoming request IP (let's call it `∑`) in the x-forwarded-for (XFF) header, when sending requests to Accedo One. The value of `∑` is what Express decides is the user’s IP.

1/ Express just acts normally when no trust proxy is set, using the incoming connection IP. So it sets `∑` to the proper IP on direct user access, and the proxy (or CDN, or such)'s IP when the user went through that.

You want to avoid reporting any proxy's IP, because it would prevent Accedo One's analytics and geolocation features from working reliably.

2/ When setting trust proxy to true (or enabling it, which does the same thing), the behaviour changes.
Now, the value of `∑` will be the first value given in the incoming request's XFF header. “First value” is the left-most value.
So what we are saying is that we trust every single proxy, and whatever ends up in XFF is trusted, so we pick the value that was first set (which is the user's IP when everybody plays nice and respects that order).
Also note, if there is no XFF header, then the connection’s IP is used as a fallback (that’s good ! it means we can handle both CDN and direct connections).

3/ When setting trust proxy to a CSV or array of IPs, those IPs are the proxies we trust. So Express looks at the XFF's right-most IP that does not match any of those trusted proxy: this would be the first “untrusted” IP, and we consider that anything that comes before that is garbage. In that case, the right-most untrusted IP is used as the value of `∑`. Here again, if there’s no XFF, the connection IP is used.

In conclusion, we recommend setting the `trust proxy` value to trust `loopback`, `linklocal`, `uniquelocal`, **AND** all IPs that are part of your proxying / CDN network ahead of the server running accedo-one-express.
Depending on your infrastructure, it may sometimes be more suitable to just set the trust proxy setting to true, or to use a custom function to recover the proper IP with a custom strategy.

## Unit Tests

Mocha (with Chai) unit tests have been written to cover all of the exported APIs from this module. Follow the following steps in order to run them:

  * Follow the **Getting Started** steps above.
  * Run `npm test`

## License

See the [LICENSE file](./LICENSE.md) for license rights and limitations (Apache 2.0)


CHANGELOG.md
# 3.0.0

- BREAKING: New name for this library and the main dependency under GitHub and NPM
- Now using Prettier to style the code

:sparkles: This is the first release published in NPM as `@accedo/accedo-one-express`. It replaces `appgrid-express`.

Accedo One is the new product that replaces AppGrid, and will bring additional possibilities

This library is now depending on `@accedo/accedo-one`, which replaces the obsolete `appgrid`.

We expect this major update to be painless for most people - just make sure to use the `res.locals.accedoOne` variable instead of `res.locals.appgrid` on browsers.

# 2.0.0

TLDR: if you upgrade from `appgrid-express` v1, just make sure to run `npm i -S appgrid` or `yarn add appgrid`.

- BREAKING: `appgrid` is moved from `dependencies` to `devDependencies` and `peerDependencies`.
Users still do not need to import (or require) `appgrid`, but they now must add it to their own project's dependencies.

This small hassle has a larger benefit: NPM and Yarn will be able to tell users when a new version of `appgrid` is available, and to update it independently. Previously, you could miss important bugfixes in `appgrid`.

# 1.1.0

- FEATURE: Add properties to the single exported function.
Those give you access to defaults used, such as the names of the default cookies, or the default `getRequestInfo` function.
This way you can decide to rely on the default `getRequestInfo` but override just the behavior to extract the `deviceId`, for instance.
See what is exported in the docs.

# 1.0.1

- FIX: Update the `appgrid` dependency to version `^3.1.4` as this fixes [an important issue](https://github.com/Accedo-Products/appgrid-sdk-js/blob/master/CHANGELOG.md#314)

# 1.0.0

- First version, extracted from the [AppGrid JS SDK](https://github.com/Accedo-Products/appgrid-sdk-js) v2.3.0 as it is now used for both Node.js and browsers.


test/specs/express.spec.js
/* eslint-disable no-unused-expressions */

const chai = require('chai');
const sinon = require('sinon');
const middleware = require('../../src/index');

const should = chai.should();

const NOOP = () => {};
const SOME_IP = '1.3.3.7';

const appKey = '56ea6a370db1bf032c9df5cb';

const buildRequest = (cookies = {}, query = {}) => ({
  cookies,
  query,
  ip: SOME_IP,
});

const buildResponse = () => ({
  locals: {},
  cookie: sinon.spy(),
});

describe('The Express middleware, with basic options', () => {
  const oneMw = middleware({
    appKey,
    gid: 'myGID',
  });

  it('exposes a function', () => {
    middleware.should.be.a('function');
  });

  it('also exposes some defaults as properties of that function', () => {
    middleware.defaultOnDeviceIdGenerated.should.be.a('function');
    middleware.defaultGetRequestInfo.should.be.a('function');
    middleware.defaultOnSessionKeyChanged.should.be.a('function');
    middleware.COOKIE_DEVICE_ID.should.be.a('string');
    middleware.COOKIE_SESSION_KEY.should.be.a('string');
  });

  it('should call the next handler in the Express pipeline', () => {
    const next = sinon.spy();
    oneMw(buildRequest(), buildResponse(), next);
    next.calledOnce.should.be.true;
  });

  it('should attach a configured locals.accedoOneClient response prop as per the parameters and cookies', () => {
    const response = buildResponse();
    const cookies = {
      [middleware.COOKIE_DEVICE_ID]: 'gregTestingSDK',
      [middleware.COOKIE_SESSION_KEY]: 'gregSessionKey',
    };
    oneMw(buildRequest(cookies), response, NOOP);
    const { accedoOneClient } = response.locals;
    accedoOneClient.should.be.an('object');
    accedoOneClient.props.config.appKey.should.equal(appKey);
    accedoOneClient.props.config.gid.should.equal('myGID');
    accedoOneClient.props.config.deviceId.should.equal(
      cookies[middleware.COOKIE_DEVICE_ID]
    );
    accedoOneClient.props.config.sessionKey.should.equal(
      cookies[middleware.COOKIE_SESSION_KEY]
    );
  });

  it('should use the gid from the request if any was found there (even if a gid was set in the middleware config)', () => {
    const response = buildResponse();
    const cookies = {
      [middleware.COOKIE_DEVICE_ID]: 'gregTestingSDK',
      [middleware.COOKIE_SESSION_KEY]: 'gregSessionKey',
    };
    const REQUEST_GID = 'a_cool_gid';
    const query = {
      gid: REQUEST_GID,
    };
    oneMw(buildRequest(cookies, query), response, NOOP);
    const { accedoOneClient } = response.locals;
    accedoOneClient.props.config.gid.should.equal(REQUEST_GID);
  });

  describe('when no deviceId or sessionKey is found in the request cookies', () => {
    const mw = middleware({
      // eslint-disable-line no-shadow
      appKey,
    });
    const response = buildResponse();
    mw(buildRequest(), response, NOOP);
    const { accedoOneClient } = response.locals;

    it('generates the deviceId on invocation and sets the deviceId response cookie', () => {
      const { deviceId } = accedoOneClient.props.config;
      deviceId.should.be.a('string');
      response.cookie.calledWith(middleware.COOKIE_DEVICE_ID, deviceId).should
        .be.true;
    });

    it('does not create a session if not needed', () => {
      const { sessionKey } = accedoOneClient.props.config;
      should.equal(false, !!sessionKey);
      response.cookie.calledWith(middleware.COOKIE_SESSION_KEY).should.be.false;
    });

    it('when a sessionKey is (re)created, sets the sessionKey response cookie', () => {
      return accedoOneClient.getApplicationStatus().then(() => {
        const { sessionKey } = accedoOneClient.props.config;
        sessionKey.should.be.a('string');
        response.cookie.calledWith(middleware.COOKIE_SESSION_KEY, sessionKey)
          .should.be.true;
      });
    });
  });
});

describe('The Express middleware used with the getRequestInfo option', () => {
  describe('when getRequestInfo returns en empty object', () => {
    const mw = middleware({
      appKey,
      gid: 'myGID',
      getRequestInfo: () => ({}),
    });

    it('should ignore the cookies used by the default getRequestInfo', () => {
      const response = buildResponse();
      const cookies = {
        [middleware.COOKIE_DEVICE_ID]: 'gregTestingSDK',
        [middleware.COOKIE_SESSION_KEY]: 'gregSessionKey',
      };
      mw(buildRequest(cookies), response, NOOP);
      const { accedoOneClient } = response.locals;
      accedoOneClient.should.be.an('object');
      // auto-generated deviceId
      accedoOneClient.props.config.deviceId.should.be.a.string;
      accedoOneClient.props.config.deviceId.should.not.equal(
        cookies[middleware.COOKIE_DEVICE_ID]
      );
      // no sessionKey
      should.equal(false, !!accedoOneClient.props.config.sessionKey);
      // gid is set as per the param
      accedoOneClient.props.config.gid.should.equal('myGID');
    });
  });

  describe('when getRequestInfo returns an object with sessionKey', () => {
    const sessionKey = 'MySessionKey';
    const mw = middleware({
      appKey,
      gid: 'myGID',
      getRequestInfo: () => ({
        sessionKey,
      }),
    });

    it('should ignore the cookies used by the default getRequestInfo, NOT use the given sessionKey because the deviceId is missing', () => {
      const response = buildResponse();
      const cookies = {
        [middleware.COOKIE_DEVICE_ID]: 'gregTestingSDK',
        [middleware.COOKIE_SESSION_KEY]: 'gregSessionKey',
      };
      mw(buildRequest(cookies), response, NOOP);
      const { accedoOneClient } = response.locals;
      // deviceId not as per the cookie
      accedoOneClient.props.config.deviceId.should.not.equal(
        cookies[middleware.COOKIE_DEVICE_ID]
      );
      // sessionKey NOT set because it's reset when there's no deviceId reused
      should.equal(false, !!accedoOneClient.props.config.sessionKey);
    });
  });

  describe('when getRequestInfo returns an object with sessionKey and deviceId', () => {
    const sessionKey = 'MyOtherSessionKey';
    const deviceId = 'ABoringDeviceId';
    const gid = 'myGID';
    const mw = middleware({
      appKey,
      gid,
      getRequestInfo: () => ({
        sessionKey,
        deviceId,
      }),
    });

    const response = buildResponse();
    const cookies = {
      [middleware.COOKIE_DEVICE_ID]: 'gregTestingSDK',
      [middleware.COOKIE_SESSION_KEY]: 'gregSessionKey',
    };
    mw(buildRequest(cookies), response, NOOP);
    const { accedoOneClient } = response.locals;

    it('should ignore the cookies used by the default getRequestInfo, use the given sessionKey', () => {
      // deviceId as given (not as per the cookie)
      accedoOneClient.props.config.deviceId.should.equal(deviceId);
      // sessionKey as given because a deviceId was also given
      accedoOneClient.props.config.sessionKey.should.equal(sessionKey);
    });

    it('should keep the gid given in the middleware constructor', () => {
      accedoOneClient.props.config.gid.should.equal(gid);
    });
  });
});


src/index.js
const accedoOne = require('@accedo/accedo-one');
const cookieParser = require('cookie-parser')();

const SIXTY_YEARS_IN_MS = 2147483647000;
const COOKIE_DEVICE_ID = 'ag_d';
const COOKIE_SESSION_KEY = 'ag_s';

// default functions for the cookie persistency strategy (deviceId and sessionKey) and gid passed as a query param
const defaultGetRequestInfo = req => ({
  deviceId: req.cookies[COOKIE_DEVICE_ID],
  sessionKey: req.cookies[COOKIE_SESSION_KEY],
  gid: req.query.gid,
});
const defaultOnDeviceIdGenerated = (id, res) =>
  res.cookie(COOKIE_DEVICE_ID, id, {
    maxAge: SIXTY_YEARS_IN_MS,
    httpOnly: true,
  });
const defaultOnSessionKeyChanged = (key, res) =>
  res.cookie(COOKIE_SESSION_KEY, key, {
    maxAge: SIXTY_YEARS_IN_MS,
    httpOnly: true,
  });

/**
 * An express-compatible middleware. This is the only export on this library.
 *
 * It's a function but has some properties attached to it that let you access some defaults:
 *
 * - 3 functions are the default ones for optional parameters detailed below:
 * `defaultOnDeviceIdGenerated`, `defaultGetRequestInfo`, `defaultOnSessionKeyChanged`
 * - 2 strings are the default cookie names:
 * `COOKIE_DEVICE_ID`, `COOKIE_SESSION_KEY`
 *
 * It uses the cookie-parser middleware, so you can also take advantage of the `req.cookies` array.
 *
 * This middleware takes care of creating an Accedo One client instance for each request automatically.
 * By default, it will also reuse and persist the deviceId and sessionKey using the request and response cookies.
 * That strategy can be changed by passing the optional callbacks,
 * so you could make use of headers or request parameters for instance.
 *
 * Each instance is attached to the response object and available to the next express handlers as `res.locals.accedoOneClient`.
 *
 * Note any extra argument provided in the config object will be passed onto the Accedo One client factory during instanciation.
 *
 * @function
 * @param  {object} config the configuration
 * @param  {string} config.appKey the application Key that will be used for all Accedo One clients
 * @param  {function} [config.getRequestInfo] callback that receives the request and returns an object with optional deviceId, sessionKey and gid properties.
 * @param  {function} [config.onDeviceIdGenerated] callback that receives the new deviceId (if one was not returned by getRequestInfo) and the response
 * @param  {function} [config.onSessionKeyChanged] callback that receives the new sessionKey (anytime a new one gets generated) and the response
 * @param  {any} [config.log/gid/etc] You can also pass any extra option accepted by the accedoOne factory function (log, gid, ...)
 * @return {function} a middleware function compatible with express
 * @example <caption>Using the default cookie strategy</caption>
 * const accedoOne = require('@accedo/accedo-one-express');
 * const express = require('express');
 *
 * const PORT = 3000;
 *
 * express()
 * // handle proxy servers if needed, to pass the user's IP instead of the proxy's.
 * .set('trust proxy', ['loopback', 'linklocal', 'uniquelocal'])
 * // place the accedoOne middleware before your request handlers
 * .use(accedoOne({ appKey: '56ea6a370db1bf032c9df5cb' }))
 * .get('/test', (req, res) => {
 *    // access your client instance, it's already linked to the deviceId and sessionKey via cookies
 *    res.locals.accedoOneClient.getEntryById('56ea7bd6935f75032a2fd431')
 *    .then(entry => res.send(entry))
 *    .catch(err => res.status(500).send('Failed to get the result'));
 * })
 * .listen(PORT, () => console.log(`Server is on ! Try http://localhost:${PORT}/test`));
 *
 * @example <caption>Using custom headers to extract deviceId and sessionKey and to pass down any change</caption>
 * const accedoOne = require('@accedo/accedo-one-express');
 * const express = require('express');
 *
 * const PORT = 3000;
 * const HEADER_DEVICE_ID = 'X-AG-DEVICE-ID';
 * const HEADER_SESSION_KEY = 'X-AG-SESSION-KEY';
 * const HEADER_GID = 'X-AG-GID';
 *
 * express()
 * .use(accedoOne({
 *   appKey: '56ea6a370db1bf032c9df5cb',
 *   // extract deviceId, sessionKey and gid from custom headers
 *   getRequestInfo: req => ({ deviceId: req.get(HEADER_DEVICE_ID), sessionKey: req.get(HEADER_SESSION_KEY), gid: req.get(HEADER_GID) }),
 *   // pass down any change on the deviceId (the header won't be set if unchanged compared to the value in getRequestInfo)
 *   onDeviceIdGenerated: (id, res) => res.set(HEADER_DEVICE_ID, id),
 *   // pass down any change on the sessionKey (the header won't be set if unchanged compared to the value in getRequestInfo)
 *   onSessionKeyChanged: (key, res) => res.set(HEADER_SESSION_KEY, key),
 *   log(...args) { console.log(...args) }
 * }))
 *
 * @example <caption>Using getRequestInfo and the defaultGetRequestInfo together</caption>
 * const accedoOne = require('@accedo/accedo-one-express');
 * const express = require('express');
 *
 * const PORT = 3000;
 * const HEADER_GID = 'X-AG-GID';
 *
 * express()
 * .use(accedoOne({
 *   appKey: '56ea6a370db1bf032c9df5cb',
 *   // extract the gid from custom headers, but use the default strategy for the deviceId and sessionKey
 *   getRequestInfo: req => Object.assign(accedoOne.defaultGetRequestInfo(req), { gid: req.get(HEADER_GID) }),
 *   log(...args) { console.log(...args) }
 * }))
 */
const accedoOneExpress = config => {
  const {
    getRequestInfo = defaultGetRequestInfo,
    onDeviceIdGenerated = defaultOnDeviceIdGenerated,
    onSessionKeyChanged = defaultOnSessionKeyChanged,
  } = config;
  return (req, res, next) =>
    cookieParser(req, res, () => {
      const { deviceId, sessionKey, gid } = getRequestInfo(req);
      const clientOptions = {
        deviceId,
        sessionKey,
        ip: req.ip,
        onDeviceIdGenerated: id => onDeviceIdGenerated(id, res),
        onSessionKeyChanged: key => onSessionKeyChanged(key, res),
      };
      // Add the gid if it was found by getRequestInfo, and not an empty string (otherwise, we will use the one passed in the config if any)
      if (gid) {
        clientOptions.gid = gid;
      }
      // Let anything given in config pass through as a client option as well
      const client = accedoOne(Object.assign({}, config, clientOptions));
      // res.locals is a good place to store response-scoped data
      res.locals.accedoOneClient = client;
      next();
    });
};

// Export useful defaults so users can wrap them, refer to them, etc
// This is done as members of the `accedoOneExpress` exported function so it's not a breaking change
Object.assign(accedoOneExpress, {
  defaultOnDeviceIdGenerated,
  defaultGetRequestInfo,
  defaultOnSessionKeyChanged,
  COOKIE_DEVICE_ID,
  COOKIE_SESSION_KEY,
});

module.exports = accedoOneExpress;
